documentation/design-docs/diagnostics-client-library.md

```diff
@@ -0,0 +1,296 @@
+
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    public class VersionInfo
+    {
+        public static const int MIN_SUPPORTED_VERSION;
+        public static const int MAX_SUPPORTED_VERSION;
+    }
+}
+```
+
+### DiagnosticsCommand
+This is an enum for all the diagnostics command supported in the current version of the library.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    /// <summary>
+    /// An enum for all the supported diagnostics IPC command
+    /// </summary>
+    public enum DiagnosticsCommand
+    {
+        Dump,
+        EventPipe
+        Profiler
+    }
+}
+```
+
+### DiagnosticsIPCHelper
+This is a class that contains some utility methods for various purposes.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    /// <summary>
+    /// A utility class that contain various helper methods to interact with the diagnostics IPC.
+    /// </summary>
+    public class DiagnosticsIPCHelper
+    {
+        /// <summary>
+        /// Get all the active processes that can be attached to.
+        /// </summary>
+        /// <returns>
+        /// IEnumerable of all the active process IDs.
+        /// </returns>
+        public static IEnumerable<int> GetActiveProcesses();
+
+        // The default path that the diagnostics client library looks for the IPC socket (default: /tmp)
+        public static string DefaultIPCSocketPath { get; set; } 
+
+        /// <summary>
+        /// Get all the supported commands by this version of the diagnostics IPC.
+        /// </summary>
+        /// <returns>
+        /// IEnumerable of all supported commands Command Code.
+        /// </returns>
+        public static IEnumerable<int> GetAvailableCommands();
+
+        /// <summary>
+        /// Convert a command name in the right format (CommandSet:Command) - i.e. EventPipe.CollectTracing - to corresponding command code.
+        /// </summary>
+        /// <returns>
+        /// Corresponding command code for the given command name. This may throw UnknownCommandException.
+        /// </returns>
+        public static int GetCommandCodeFromCommandName(string commandName);
+
+        /// <summary>
+        /// Convert a command code to its string representation
+        /// </summary>
+        /// <returns>
+        /// Corresponding command name for the given command code. This may throw UnknownCommandException
+        /// </returns>
+        public static string GetCommandNameFromCommandCode(int commandCode);
+
+    }
+}
+```
+
+### EventPipeCommandHandler
+This is a CommandHandler class for sending EventPipe commands across the diagnostics IPC channel.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    /// <summary>
+    /// A class for handling EventPipe commands.
+    /// </summary>
+    public class EventPipeCommandHandler
```

Comment:
Consumers of this class will have a hard time unit testing. Consider adding an interface or making methods virtual.

Reply:
Thanks for the comments @loic-sharma - I'll dig on this one for a little bit before making changes. I'm leaning towards adding an interface for now though.

Reply:
From my understanding, default interface methods were added to mitigate this exact problem. See: https://github.com/dotnet/corefx/pull/41949/files

Reply:
My understanding is the default interface method needs to be implementable in terms of the other methods already on the interface. That doesn't seem feasible here?